File: client/lib/core/models/task.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:uuid/uuid.dart';

part 'task.g.dart';

enum TaskType {
  @JsonValue('RESEARCH')
  research,
  @JsonValue('STRATEGY_DEV')
  strategyDev,
  @JsonValue('BACKTEST')
  backtest,
}

enum TaskStatus {
  @JsonValue('PLANNING')
  planning,
  @JsonValue('PENDING_APPROVAL')
  pendingApproval,
  @JsonValue('RUNNING')
  running,
  @JsonValue('COMPLETED')
  completed,
  @JsonValue('FAILED')
  failed,
}

@JsonSerializable()
class Task {
  final String id;
  final String title;
  final String? description;
  final TaskType taskType;
  final TaskStatus status;
  final Map<String, dynamic> inputData;
  final Map<String, dynamic>? generatedPlan;
  final Map<String, dynamic>? result;
  final String? errorDetails;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? dueDate;

  Task({
    required this.id,
    required this.title,
    this.description,
    required this.taskType,
    required this.status,
    required this.inputData,
    this.generatedPlan,
    this.result,
    this.errorDetails,
    required this.createdAt,
    required this.updatedAt,
    this.dueDate,
  });

  factory Task.fromJson(Map<String, dynamic> json) => _$TaskFromJson(json);
  Map<String, dynamic> toJson() => _$TaskToJson(this);

  Task copyWith({
    String? id,
    String? title,
    String? description,
    TaskType? taskType,
    TaskStatus? status,
    Map<String, dynamic>? inputData,
    Map<String, dynamic>? generatedPlan,
    Map<String, dynamic>? result,
    String? errorDetails,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? dueDate,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      taskType: taskType ?? this.taskType,
      status: status ?? this.status,
      inputData: inputData ?? this.inputData,
      generatedPlan: generatedPlan ?? this.generatedPlan,
      result: result ?? this.result,
      errorDetails: errorDetails ?? this.errorDetails,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      dueDate: dueDate ?? this.dueDate,
    );
  }
}

@JsonSerializable()
class TaskCreate {
  final String title;
  final String? description;
  final TaskType taskType;
  final Map<String, dynamic> inputData;
  final DateTime? dueDate;

  TaskCreate({
    required this.title,
    this.description,
    required this.taskType,
    required this.inputData,
    this.dueDate,
  });

  factory TaskCreate.fromJson(Map<String, dynamic> json) => _$TaskCreateFromJson(json);
  Map<String, dynamic> toJson() => _$TaskCreateToJson(this);
}

@JsonSerializable()
class PlanApproval {
  final bool approved;
  final Map<String, dynamic>? modifiedPlan;

  PlanApproval({
    required this.approved,
    this.modifiedPlan,
  });

  factory PlanApproval.fromJson(Map<String, dynamic> json) => _$PlanApprovalFromJson(json);
  Map<String, dynamic> toJson() => _$PlanApprovalToJson(this);
} 

File: app/models/task.py
from enum import Enum as PyEnum
import uuid
from sqlalchemy import Column, String, JSON, Text, DateTime, Enum
from sqlalchemy.dialects.postgresql import UUID, JSONB
from datetime import datetime
from app.db.session import Base

class TaskType(str, PyEnum):
    RESEARCH = "RESEARCH"
    STRATEGY_DEV = "STRATEGY_DEV"
    BACKTEST = "BACKTEST"

class TaskStatus(str, PyEnum):
    PLANNING = "PLANNING"
    PENDING_APPROVAL = "PENDING_APPROVAL"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"

class Task(Base):
    __tablename__ = "tasks"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    task_type = Column(Enum(TaskType), nullable=False)
    status = Column(Enum(TaskStatus), nullable=False, default=TaskStatus.PLANNING)
    input_data = Column(JSONB, nullable=False)
    generated_plan = Column(JSONB)
    result = Column(JSONB)
    metadata = Column(JSONB, default={})
    
    __table_args__ = (
        Index('idx_task_type_status', 'task_type', 'status'),
        Index('idx_created_at', 'created_at'),
    )
    error_details = Column(Text)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)